DOCSP-45307: Drop external polyfill library support (#74)
mongodb/docs-cpp: source/index.txt

Welcome to the documentation site for the official {+driver-long+}.

C++17 Polyfill Configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The {+driver-long+} uses C++17 features ``std::optional<T>`` and ``std::string_view``.
If you configure the driver with a pre-C++17 standard, the bsoncxx library provides
polyfill implementations for these C++17 features. To limit reliance on polyfill-specific
behavior, avoid using ``stdx::optional<T>`` and ``stdx::string_view`` for pre-C++17 configurations.
The driver uses the bsoncxx polyfill implementations when ``CMAKE_CXX_STANDARD`` is less than ``17``
(set to ``11`` by default).

We recommend using the C++17 standard library implementations whenever possible. To use

> these implementations, set ``CMAKE_CXX_STANDARD`` to ``17`` during CMake configuration.

Author: I added back the admonition information. Just to clarify: what is "non-bsoncxx and non-mongocxx library interfaces" referring to, if the driver drops support for all external polyfill implementations & automatically uses bsoncxx for pre-C++17 configurations? Is it still relevant to refer to a "choice of polyfill library"?

> these implementations, set ``CMAKE_CXX_STANDARD`` to ``17`` during CMake configuration.

Reviewer: It refers to other libraries or user-defined code which a user might be using alongside the C++ Driver. For example, a user compiling against the C++ Driver libraries configured to use the Boost library as polyfill might also be using Boost libraries themselves and expect our polyfills to be equivalent to the Boost library interfaces.

```cpp
// C++ Driver interfaces.
bsoncxx::stdx::string_view foo();
bsoncxx::stdx::optional<T> bar();

// Before (configured with Boost as polyfill library):
boost::string_view sv = foo();  // OK: same type.
boost::optional<T> opt = bar(); // OK: same type.

// After (support for Boost as polyfill library is removed):
auto bsv = foo();  // bsoncxx::stdx::string_view is not boost::string_view!
auto bopt = bar(); // bsoncxx::stdx::optional<T> is not boost::optional<T>!

// Required API breaking changes in user code.
boost::string_view sv{bsv.data(), bsv.size()};      // Translate from stdx to Boost.
boost::optional<T> opt{bopt ? *bopt : boost::none}; // Translate from stdx to Boost.
```

This principle still applies, but now for the bsoncxx polyfill implementation instead of Boost or mnmlstc/core. This is especially so for users who set `BSONCXX_POLY_USE_IMPLS=ON` (overriding default polyfill library selection behavior which would use the standard library with C++17 or newer) or are upgrading from pre-C++17 to C++17. Even if we do our best to ensure compatibility with their stdlib counterparts, they _aren't_ the stdlib counterparts themselves, so there is some inevitable opportunity for inconsistency, which is what the final sentence of the warning is noting.

> these implementations, set ``CMAKE_CXX_STANDARD`` to ``17`` during CMake configuration.

Author: I see, thanks for the explanation!

Welcome to the documentation site for the official {+driver-long+}.

C++17 Polyfill Configuration
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The {+driver-long+} uses C++17 features ``std::optional<T>`` and ``std::string_view``.
If you configure the driver with a pre-C++17 standard, the bsoncxx library provides
polyfill implementations for these C++17 features. The driver uses the bsoncxx polyfill
implementations when the ``CMAKE_CXX_STANDARD`` configuration option is set to a number
less than ``17``. By default, this option is set to ``11``.

> C++ standard conformance and supported behavior of polyfill features might vary depending
> on the selected polyfill library. The purpose of these polyfills is to support pre-C++17
> configurations by providing stand-ins for their C++17 equivalents. We recommend
> using the C++17 standard library whenever possible by setting the ``CMAKE_CXX_STANDARD``
> configuration option to ``17`` or newer.

Reviewer: ```suggestion
We recommend using the C++ standard library instead of polyfills by setting the ``CMAKE_CXX_STANDARD`` configuration option to ``17`` or newer.
```
Suggest removing the first two sentences and making only the last sentence an admonition to emphasize the importance of the "please use C++17 or newer" recommendation, e.g.:

> [!IMPORTANT]
> We recommend using the C++ standard library instead of bsoncxx polyfill implementations by setting the ``CMAKE_CXX_STANDARD`` configuration option to ``17`` or newer.

Author: I moved this admonition to the beginning, just to avoid stacking admonitions
